### ISSUE TITLE
Feature request: `npm pack` command equivalent

### ISSUE THREAD

**Issue by trevor-scheer** [changesets/changesets#728]
The [`npm pack`](https://docs.npmjs.com/cli/v8/commands/npm-pack) command allows one to create a tarball from an npm package. This tarball can then be `npm install`ed into a project (typically for local development, testing, early builds).

## Desired outcome

A `changeset pack` command which creates tarballs for each package in the monorepo.

---

## Our use case
At Apollo, we currently accomplish this in a bit of a roundabout way on a couple of our monorepos (which currently use Lerna for the release process). To do this, we actually perform a lerna publish, but capture the results in a Verdaccio registry instead of publishing to NPM. We then upload the tarballs in Circle CI as an artifact with an `npm install`able url. This allows us to share early builds directly from a PR.

Here's an example of a particular build's artifacts for Apollo's federation repo (these artifacts will eventually expire):
https://app.circleci.com/pipelines/github/apollographql/federation/6561/workflows/a779c552-df37-49c8-bf06-3bc5f6b1c64a/jobs/38049/artifacts

If I copy link address for a tarball, I can just:
npm i https://38049-252760990-gh.circle-artifacts.com/0/storage/%40apollo/federation/federation-0.33.8.tgz
...and try out / share this particular build of the federation package.

I'm currently exploring `changesets` for a greenfield project, and would be interested in bringing this same functionality to the new project - ideally as an actual command within `changesets`.

---

### Problem: packing dependent packages

Packing interdependent packages (A depends on B) has some challenges. If you think about A's `package.json`, the final packed output for A will naturally depend on the `latest` _published_ version of B. If you install the packed 'A', you will end up with the local version of A and the registry version of B. Installing a packed 'A' should typically _work_, assuming B has been published at least once. However, if B also has changes, those won't be pulled in when A is installed (since it depends on the latest/registry version of B).

#### Solutions?
* This can be worked around by installing the packed version of B into your project as well (where you otherwise might not have had to). This isn't ideal for a developer workflow, but it's also an acceptable hiccup (at least it has been for us, this is what we currently do).
* Allow the user to specify a way to resolve interdependent packages' versions as an input to the pack command. I think this is ideal - if I can determine the url (or just location on disk) where a tarball of B will exist ahead of time, we can replace the version with the link/location in A's `package.json`. This at least addresses the drawback to _our_ existing implementation. Since I haven't actually been able to do this in practice, I'm not sure if it has any real drawback but it seems pretty sound in theory.

**Comment by trevor-scheer** [comment]
So I peeked at a PR in this repo and it looks like the codesandbox bot does exactly what I'm proposing, down to the second solutions bullet I hypothesized. Feel free to leave this open if you think `changesets` should support this natively. Otherwise, we can close this since it's a solved problem.

**Comment by Andarist** [comment]
The feature itself sounds nice - but it comes with this exact gotcha that you have mentioned. The problem is that it's all depending on how do you host those "snapshot" artifacts - and, like you have observed, CodeSandbox CI solves this quite nice because it's all neatly automated, they already understand the dependency graph, they are able to pack things correctly **while** rewriting declared dependencies with links pointing to their "registry".

There are two problems with CodeSandbox CI though:
- IIRC this only works (at least atm) with public repos
- they rewrite deps in a way that is not compatible with Yarn Berry, I've described the issue [here](https://github.com/codesandbox/codesandbox-ci-ui/pull/34#issuecomment-995829289) and figured out a workaround that can be used to fix this [here](https://github.com/yarnpkg/berry/issues/3327#issuecomment-995886724)

Even despite those (somewhat minor) problems, I believe there might be still a need for a solution that would be implemented in Changesets because it would allow for this to be used in more contexts, custom registries, artifactories etc. As mentioned though, we don't know where one is going to "publish" their packed stuff so I imagine that we'd have to accept a pattern that would be used to rewrite dependencies, so smth like this could potentially work:
```
changeset pack --rewrite-with=https://38049-252760990-gh.circle-artifacts.com/0/storage/%40apollo/federation/{tarball}
```
Where `{tarball}` would be replaced with the appropriate filename for a given package. Maybe we could simplify this and just accept a URL and assume that the tarball's filename can be put at the end.

That being said - I don't think I will have time to work on implementing such a feature any time soon. If this is important for anyone - I would gladly discuss the implementation details and accept a PR for this.

**Comment by trevor-scheer** [comment]
Sounds good! I agree it seems to have a place here. I'd only add that supporting file paths as well would be a nice addition for the local development use case (and I doubt this adds much if any complexity to the solution).

The `--rewrite-with` option seems sufficient to me. Dynamic requirements could be satisfied via env variables.

I no longer need this functionality but that could change in the future. In any case, I'm happy to participate on this issue or if a PR is submitted to address it.

Thanks for your input!

**Comment by Andarist** [comment]
>  I'd only add that supporting file paths as well would be a nice addition for the local development use case (and I doubt this adds much if any complexity to the solution).

How this would work though? Would absolute paths be inserted as dependencies?

**Comment by trevor-scheer** [comment]
Hm, I suppose this does get brittle with respect to moving these tarballs around.

[`bundledDependencies`](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#bundleddependencies) might have a place in solving this. If the packages are bundled topographically and `package.json`s rewritten appropriately this could be a solid approach. It would pack each tarball with its own local copy of cross-dependencies, making the output tarballs completely independent of eachother.

**Comment by Andarist** [comment]
I'd have to read more about how `bundledDependencies` work. While I know that they exist - I have never utilized them. My suspicion though is that they prevent sharing the dependency between multiple packages.

So let's say that you have packages A and B, both depending on C and you install both A and B with C being their bundled dependency. If this results in C being isolated for each dependent package then we can run into all sorts of problems: `instanceof` might not work properly, React context might not work properly, etc.

**Comment by alexandcote** [comment]
👋🏻 How can I help implement a `changeset pack` command 😄 

Our use case is a little bit simpler, we simply want to generate the tarball and manually upload them to our private registry. What would be the preferred approach here? I feel like the `--rewrite-with` option can be a follow-up.

**Comment by trevor-scheer** [comment]
@alexandcote if you're dealing with a monorepo you're going to run into the mentioned gotcha without the `--rewrite-with` option. If you're not in a monorepo you should just be able to `npm pack` to get the tarball you need.

**Comment by Andarist** [comment]
@alexandcote why can't you use `changeset version && npm/yarn/pnpm pack`? `changeset publish` doesn't do a lot of stuff under the hood, most of what it does is related to publishing to a registry. So I just wonder what you expect that said `chngeset pack` to do cause at the moment I see it mostly as a proxy to the underlying `npm/yarn/pnpm pack`.

**Comment by Andarist** [comment]
> This is what we are doing at the moment but from my understanding, in a monorepo setup, you have to loop across different packages and run npm/yarn/pnpm pack in each package.

Yeah, that's likely true - unless you have an access to a package manager that can do it for you. For example, it's possible to do `pnpm --recursive publish` but, from what I can read in their docs, this isn't supported for all commands so there is no `pnpm --recursive pack`.

Many of the existing task runners (like Lerna, Nx, Turborepo, etc) allow you to execute a command in each workspace. Alternatively, you can write a custom script that loops through all of the packages in the monorepo and calls that command there. You can get the list of packages with `@manypkg/get-packages`.

> I can definitely create a node/bash script doing that but I thought it could be nice to implement this logic directly in the lib so each consumer doesn't have to maintain their custom script.

Yeah, I'm not saying no to that - but for the most part, it feels easily achievable in the userland. And we would have to establish what this command should do exactly, what kind of options it has to support, etc. It just seems like so thin layer over the existing tools that it's on the verge of "not worth it".

That being said, from what I understand - you are willing to work on this. So you have to implement this for your project anyway and with that work being done, you could try to port it to Changesets in a form of a PR so I could see what would be the exact complexity of the proposed feature.

**Comment by bluwy** [comment]
ref: https://github.com/changesets/changesets/issues/1780

**Comment by bluwy** [comment]
Resolved in https://github.com/changesets/changesets/pull/2062
